addons/mail/models/discuss/discuss_channel.py

```diff
@@ -323,30 +328,38 @@ def _subscribe_users_automatically_get_members(self):
     def action_unfollow(self):
         self._action_unfollow(self.env.user.partner_id)
 
-    def _action_unfollow(self, partner):
+    def _action_unfollow(self, partner=None, guest=None):
+        self.ensure_one()
         self.message_unsubscribe(partner.ids)
         custom_store = Store(self, {"is_pinned": False, "isLocallyPinned": False})
         member = self.env["discuss.channel.member"].search(
-            [("channel_id", "=", self.id), ("partner_id", "=", partner.id)]
+            [
+                ("channel_id", "=", self.id),
+                ("partner_id", "=", partner.id) if partner else ("guest_id", "=", guest.id),
+            ]
         )
         if not member:
-            partner._bus_send_store(custom_store, notification_type="discuss.channel/leave")
+            target = partner or guest
+            target._bus_send_store(custom_store, notification_type="discuss.channel/leave")
             return
-        member.unlink()
-        notification = Markup('<div class="o_mail_notification">%s</div>') % _("left the channel")
-        # sudo: mail.message - post as sudo since the user just unsubscribed from the channel
-        self.sudo().message_post(
-            body=notification, subtype_xmlid="mail.mt_comment", author_id=partner.id
-        )
+        if not member.channel_id.parent_channel_id:
+            notification = Markup('<div class="o_mail_notification">%s</div>') % _(
+                "left the channel"
+            )
+            # sudo: mail.message - post as sudo since the user just unsubscribed from the channel
+            member.channel_id.sudo().message_post(
+                body=notification, subtype_xmlid="mail.mt_comment", author_id=partner.id
+            )
         # send custom store after message_post to avoid is_pinned reset to True
-        partner._bus_send_store(custom_store, notification_type="discuss.channel/leave")
-        self._bus_send_store(
+        member._bus_send_store(custom_store, notification_type="discuss.channel/leave")
+        member.channel_id._bus_send_store(
             self,
             {
                 "channelMembers": Store.many(member, "DELETE", only_id=True),
                 "memberCount": self.member_count,
             },
         )
+        member.unlink()
```

Comment:
This can't be unlinked afterwards, as data added to store will incorrectly take the leaving member into account.


If this is green, we need new tests to cover it (`member_count` should definitely be wrong).